src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,325 @@ protected override void InsertItem(int index, IDeltaSetItem item)
+                    // The apiHandlerOfT is MyOverdueOrdersAPIHandler.
+                    // The object with id 1 will use MyOverdueOrdersAPIHandler since odata path will be MyOverdueOrders(1)
+                    // The object with odata id Employees(2)/NewFriends(2)/Microsoft.Test.E2E.AspNet.OData.BulkOperation.MyNewFriend/MyNewOrders(2)
+                    // will use NewOrdersAPIHandler.
+                    // The codebelow ensures we use the correct handler.
+
+                    IODataAPIHandler odataPathApiHandler = apiHandlerFactory.GetHandler(changedObj.ODataPath);
+
+                    if (odataPathApiHandler != null && changedObj.ODataPath.Any() && apiHandler.ToString() != odataPathApiHandler.ToString())
+                    {
+                        handler = odataPathApiHandler as ODataAPIHandler<TStructuralType>;
+                    }
+                }
+
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                Dictionary<string, object> keyValues = new Dictionary<string, object>();
+
+                foreach (string key in _keys)
+                {
+                    object value;
+
+                    if (changedObj.TryGetPropertyValue(key, out value))                   
+                    {
+                        keyValues.Add(key, value);
+                    }
+                }
+
+                try
+                {
+                    ODataAPIResponseStatus odataAPIResponseStatus = handler.TryGet(changedObj.ODataPath.GetKeys(), out original, out getErrorMessage);
+
+                    DeltaDeletedEntityObject<TStructuralType> deletedObj = changedObj as DeltaDeletedEntityObject<TStructuralType>;
+
+                    if (odataAPIResponseStatus == ODataAPIResponseStatus.Failure || (deletedObj != null && odataAPIResponseStatus == ODataAPIResponseStatus.NotFound))
+                    {
+                        IDeltaSetItem deltaSetItem = changedObj;
+                        DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+                        dataModificationExceptionType.MessageType = new MessageType { Message = getErrorMessage };
+
+                        deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation(SRResources.DataModificationException, dataModificationExceptionType);
+                        deltaSet.Add(deltaSetItem);
+                        
+                        continue;
+                    }
+
+                    if (deletedObj != null)
+                    {
+                        operation = DataModificationOperationKind.Delete;
+                        changedObj.CopyChangedValues(original, handler, apiHandlerFactory);
```

Comment:
Why do we patch the object here just to go ahead and nuke it by calling `TryDelete`? Is it to ensure that if `TryDelete` fails the object will have been patched?

Reply:
I don't see any value in patching here.

Reply:
Aah. So have seen why we patch. We patch to remove the relationship between the top-level resource and the linked resource what is pretty much what calling TryDelete will do.

Reply:
How do the `@odata.remove` values of `changed` and `deleted` get applied here.